python/dgl/graphbolt/impl/ondisk_dataset.py

```diff
@@ -38,6 +39,234 @@
+        src, dst = read_edges(dataset_dir, edge_fmt, edge_path)
+        num_nodes = graph_data["nodes"][0]["num"]
+        num_edges = len(src)
+        coo_tensor = torch.tensor([src, dst])
+        sparse_matrix = spmatrix(coo_tensor, shape=(num_nodes, num_nodes))
+        indptr, indices, value_indices = sparse_matrix.csc()
+        node_type_offset = None
+        type_per_edge = None
+        node_type_to_id = None
+        edge_type_to_id = None
+        node_attributes = {}
+        edge_attributes = {}
+        if include_original_edge_id:
+            edge_attributes[ORIGINAL_EDGE_ID] = value_indices
+    else:
+        # Heterogeneous graph.
+        # Sort graph_data by ntype/etype lexicographically to ensure ordering.
+        graph_data["nodes"].sort(key=lambda x: x["type"])
+        graph_data["edges"].sort(key=lambda x: x["type"])
+        # Construct node_type_offset and node_type_to_id.
+        node_type_offset = [0]
+        node_type_to_id = {}
+        for ntype_id, node_info in enumerate(graph_data["nodes"]):
+            node_type_to_id[node_info["type"]] = ntype_id
+            node_type_offset.append(node_type_offset[-1] + node_info["num"])
+        # Construct edge_type_offset, edge_type_to_id and coo_tensor.
+        edge_type_offset = [0]
+        edge_type_to_id = {}
+        coo_src_list = []
+        coo_dst_list = []
+        coo_etype_list = []
+        for etype_id, edge_info in enumerate(graph_data["edges"]):
+            edge_type_to_id[edge_info["type"]] = etype_id
+            edge_fmt = edge_info["format"]
+            edge_path = edge_info["path"]
+            src, dst = read_edges(dataset_dir, edge_fmt, edge_path)
+            edge_type_offset.append(edge_type_offset[-1] + len(src))
+            src_type, _, dst_type = etype_str_to_tuple(edge_info["type"])
+            src += node_type_offset[node_type_to_id[src_type]]
+            dst += node_type_offset[node_type_to_id[dst_type]]
+            coo_src_list.append(torch.tensor(src))
+            coo_dst_list.append(torch.tensor(dst))
+            coo_etype_list.append(torch.full((len(src),), etype_id))
+
+        coo_src = torch.cat(coo_src_list)
+        coo_dst = torch.cat(coo_dst_list)
+        coo_etype = torch.cat(coo_etype_list)
+
+        total_num_nodes = node_type_offset[-1]
+        total_num_edges = len(coo_src)
+
+        sparse_matrix = spmatrix(
+            indices=torch.stack((coo_src, coo_dst), dim=0),
+            shape=(total_num_nodes, total_num_nodes),
+        )
+        indptr, indices, value_indices = sparse_matrix.csc()
+        node_type_offset = torch.tensor(node_type_offset)
+        type_per_edge = torch.index_select(
+            coo_etype, dim=0, index=value_indices
+        )
+        node_attributes = {}
+        edge_attributes = {}
+        if include_original_edge_id:
+            original_edge_id = [
+                value_indices[i] - edge_type_offset[type_per_edge[i]]
+                for i in range(total_num_edges)
```

Comment:
explicit for loop could be quite inefficient. equivalent with below one which eliminate explicit loop and converting to tensor back.
```
value_indices - edge_type_offset[type_per_edge[value_indices]]
```